Better json handling and cleanup. DSDEGP-1070

kshakir opened the pull request:
### Description

Better json handling.
Using instances of enumeratum.Enum instead of String constants.
Json decoders no longer allow extra object keys from web service requests and elasticsearch responses.
Json printers no longer output scala.None as null in web service responses and elasticsearch requests.
Json printers print object keys with snake_case names.
Moved existing models to server.
Implicit parameters have full names instead of abbrs.
Renamed ElasticsearchDAO to SearchDAO.
Refactored the HttpElasticsearcDAO implementation.

The current models are server only. There are forthcoming DTOs that may be shared between the existing clio-server and a future clio-client.

DSDEGP-1070

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: Ensure that you have added or updated tests
- [x] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Squash commits and merge to develop
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [x] **Submitter**: Move the JIRA issue to QA once this checklist is completed

broadinst commented:
Can one of the admins verify this patch?

tbl3rd commented on clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/SystemElasticsearchDAO.scala:
```diff
@@ -0,0 +1,111 @@
+  override def initialize(): Future[Unit] = {
+    for {
+      _ <- waitForSearchReady()
+      _ <- createOrUpdateIndex(ElasticsearchIndex.ReadGroup)
```
Verify at startup time that all the index types supported are available?

kshakir replied:
Yep

tbl3rd commented on clio-server/src/main/scala/org/broadinstitute/clio/server/service/ServerService.scala:
```diff
@@ -62,8 +58,7 @@ class ServerService private (
       _ <- serverStatusDAO.setStatus(ServerStatusInfo.Starting)
-      _ <- waitForElasticsearchReady()
-      _ <- createOrUpdateIndices()
+      _ <- searchDAO.initialize()
```
Why isn't initialize() also named startup(), or the other way round?

kshakir replied:
The ES server must already be running, but may be empty. `initialize()` will create or update the mappings. Meanwhile the akka-http server `startup()` actually starts listening for connections to the server.

tbl3rd commented on clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/SystemElasticsearchDAO.scala:
```diff
@@ -0,0 +1,111 @@
+import scala.concurrent.Future
+import scala.util.{Failure, Success}
+
+trait SystemElasticsearchDAO extends SearchDAO { this: HttpElasticsearchDAO =>
```
I don't understand this at all.
(I don't really understand ElasticSearchContainer either.)
http://docs.scala-lang.org/tutorials/tour/self-types.html
http://www.cakesolutions.net/teamblogs/2011/12/19/cake-pattern-in-depth
http://jonasboner.com/real-world-scala-dependency-injection-di/
It allows SystemElasticsearchDAO to call getClusterHealth and so on?
How did you arrive at this design?

kshakir replied:
Hopefully discussed in person, still:

> It allows SystemElasticsearchDAO to call getClusterHealth and so on?

Yes, it specifies that an SystemElasticsearchDAO may only be used in a place where the trait is mixed in with an HttpElasticsearchDAO.

Meanwhile, HttpElasticsearchDAO mixes in SystemElasticsearchDAO, so the contract is satisfied.

> How did you arrive at this design?

I arrived at this pattern because I wanted to start separating different parts of a massive HttpElasticsearchDAO into separate traits. The underlying SearchDAO interface/trait will eventually contain dozens of abstract methods. Instead of having all the concrete methods implemented in one massive class, I'd rather have them grouped into different mixins.

H/T also to the Firecloud-Orchestration [design](https://broadinstitute.atlassian.net/wiki/spaces/DSDEGP/pages/114531509/Clio+Metadata+Manager#Clio(MetadataManager)-PackageStructureandSingleResponsibilities).

tbl3rd reviewed (approved):
Have not got through all of it yet.

kshakir commented:
Minor doc patches being squashed in, then merging after verbal ok from TBL.

kshakir committed:
Cleanup.
Implicits parameters have full names instead of abbrs.
Renamed ElasticsearchDAO to SearchDAO.

kshakir committed:
Refactored the HttpElasticsearcDAO implementation.
Search status only returns OK or Error.

kshakir committed:
Moved existing models to server.

kshakir committed:
Better json handling.
Using instances of enumeratum.Enum instead of String constants.
Json decoders no longer allow extra object keys from web service requests and elasticsearch responses.
Json printers no longer output scala.None as null in web service responses and elasticsearch requests.
Json printers print object keys with snake_case names.
